### PR TITLE
Add support for Pod Security Admission in the Namespace resource

### LINKED ISSUE #7617: Add support for Pod Security Admission in the Namespace resource
# Description

As outcome of https://github.com/rancher/dashboard/issues/7119, add mentioned features:
1. Namespace creation/edit UI
2. ~Project creation/edit UI (optional)~ Not for this initial implementation
3. Project/Namespace list
4. Namespace Detail View

## Context

The ability to define label is already allowed in the current state, without the need to have anything from the server side defined. In this way we could already proceed in integrating this functionality.

### PR DESCRIPTION
**Pull request by cnotv**
### Summary
Fixes #7617
This PR introduces Pod Security Admission (PSA) for Namespaces and affect changes in several views.

### Occurred changes and/or fixed issues
- In Namespace and Project editors, added PSA tab with smaller PSA form
- In Namespace and Project list, added icon with tooltip for PSA resources, if any
- In Namespace details, added label for PSA

### Technical notes summary

Tests involving checkbox and select input components have been avoided due complication in identify the implementation of it. Case added to the [troubleshooting list](https://github.com/rancher/dashboard/issues/7464).

### Areas or cases that should be tested
See issue for the details of each:
- Namespace CRUD

### Screenshot/Video

https://user-images.githubusercontent.com/5009481/205892111-8893ac0a-8ee1-41be-bd51-0196f9648b70.mov

Labels after PR update:

Screenshot 2022-12-16 at 14 42 42

### CODE REVIEW

File: shell/models/pod.security.admission.js
Comment (cnotv) [reply shell/models/pod.security.admission.js:1]: Not for this implementation. Removed.

File: shell/components/__tests__/PodSecurityAdmission.test.ts
Comment (cnotv) [reply shell/components/__tests__/PodSecurityAdmission.test.ts:51]: We needed as we find why it's not possible to use other inputs. Suggestions are appreciated :P

Comment (cnotv) [reply shell/components/__tests__/PodSecurityAdmission.test.ts:51]: I add a `TODO` connected to the troubleshooting.

File: shell/components/__tests__/PodSecurityAdmission.test.ts
Comment (cnotv) [reply shell/components/__tests__/PodSecurityAdmission.test.ts:73]: Added TODO.

**Review by richard-cox — CHANGES_REQUESTED** [review]
There's some places we could use the generic framework rather than specific PSA code, I need to come back a mark these out and include info how it can be done

File: shell/components/LabelsPSA.vue
Comment (richard-cox) [inline shell/components/LabelsPSA.vue:29]: This one can be moved to the model for the value type as well

File: shell/components/LabelsPSA.vue
Comment (richard-cox) [inline shell/components/LabelsPSA.vue:1]: Optional - I haven't seen this running yet, but low bar solution would be to just pass the `mode` through to the component's of the PodSecurityAdmission component. If the mode is something not editable the controls will then be disabled.

File: shell/components/PodSecurityAdmission.vue
Comment (richard-cox) [inline shell/components/PodSecurityAdmission.vue:54]: Couldn't see where this is wired in, missing a TODO or can it be removed?

File: shell/components/ResourceDetail/index.vue
Comment (richard-cox) [inline shell/components/ResourceDetail/index.vue:381]: `tooltipsDescription` should be a fn in the root `resource-class` that returned undefined (or not and we check for undefined) and have an override in the model for this resource type

```
liveModel?.detailTopToolTip

get detailTopToolTip() { ... }
```

File: shell/edit/namespace.vue
Comment (richard-cox) [inline shell/edit/namespace.vue:230]: Users should still be able to edit labels and annotations for a namespace

**Review by richard-cox — CHANGES_REQUESTED** [review]
I didn't give much of the functionality a test but did notice there's an error when looking at the detail page of a namespace that had the new lock icon.

Once review comments are addressed will double check how it all runs

File: shell/components/DetailTop.vue
Comment (cnotv) [reply shell/components/DetailTop.vue:102]: The component is full of logic on its own on manipulating the `value`, do you want to refactor it now?

File: shell/components/ExplorerProjectsNamespaces.vue
Comment (cnotv) [reply shell/components/ExplorerProjectsNamespaces.vue:245]: You may want to use the function outside of the model, so you don't have to import everything.
For instance, you are in this case into the namespace, not in the PSA. How do you deal with that?

File: shell/edit/namespace.vue
Comment (cnotv) [reply shell/edit/namespace.vue:230]: Do we have 2 types of labels then? @kwwii could you please clarify this point?

File: shell/components/ExplorerProjectsNamespaces.vue
Comment (cnotv) [reply shell/components/ExplorerProjectsNamespaces.vue:410]: Where should `row.detailLocation` be written specifically?

File: shell/components/LabelsPSA.vue
Comment (cnotv) [reply shell/components/LabelsPSA.vue:29]: How can I access to the PSA here? This resource (`value`) is a Namespace in this case, so it has not the method. Or are you asking me to write it twice for the Namespace and the PSA?

File: shell/components/LabelsPSA.vue
Comment (cnotv) [reply shell/components/LabelsPSA.vue:1]: Only some keys are disabled, not all.
All the keys are put together as all the labels are together as well.
Please refer to the design.

File: shell/components/ResourceDetail/index.vue
Comment (cnotv) [reply shell/components/ResourceDetail/index.vue:381]: This is for the Namespace.
The same function is used also later in the PSA.
We need 2 functions in 2 different models, where we are not needing the class.
Again we need just the utility function here, so there's no gain is moving it to the model.

**Comment by cnotv** [comment]
@richard-cox as help to understand the needs for the utilities, please rely on the new PR for the resource itself.

https://github.com/rancher/dashboard/pull/7678

File: shell/edit/namespace.vue
Comment (kwwii) [reply shell/edit/namespace.vue:230]: Yes, essentially this brings in a new type of label - a "system" label which can be shown or hidden by the user. A "system" label is a label that is used by Rancher itself to perform specific actions, in comparison to "user" labels which can be set by the user for use outside of the Rancher specific scope.

Comment (cnotv) [reply shell/edit/namespace.vue:230]: What is the required action then? Should we keep `Label and Admission` or not?

Comment (kwwii) [reply shell/edit/namespace.vue:230]: I assume you mean Annotations? If so, yes, there still needs to be a possibility for a user to enter a label or annotation that is not a PSA (or other possible system labels in the future).

Comment (cnotv) [reply shell/edit/namespace.vue:230]: Ok but in these `Labels and Annotation` the user will see exactly the same `Labels` and on top of it the `Annotation`, so what's the sense of our custom design?

Comment (richard-cox) [reply shell/edit/namespace.vue:230]: The `Labels and Annotations` tab should still be shown. It's `Labels` component needs updating to have a button (situational given that it's only needed for PSA world?) to reveal the system labels (PSAs)

https://github.com/rancher/dashboard/issues/7119#issuecomment-1292511032

File: shell/components/DetailTop.vue
Comment (richard-cox) [reply shell/components/DetailTop.vue:102]: The component generically handles input from `value` and converts it into a specific format that it needs to display the information. Moving type specific code into the model will maintain that generic pattern

File: shell/components/ExplorerProjectsNamespaces.vue
Comment (richard-cox) [reply shell/components/ExplorerProjectsNamespaces.vue:245]: `shell/models/namespace.js` 

```
  get hasPSALabels() {
    return hasPSALabels(row); 
  }
```

Though i think we can remove the need to use it outside of the models

File: shell/edit/namespace.vue
Comment (richard-cox) [reply shell/edit/namespace.vue:230]: Ok, i've seen it running again. The only thing visually missing here is the `Annotations`. The current `Labels` component which currently covers both labels and annotations should probably be beefed up to optionally include the `show system label` toggle.

File: shell/components/LabelsPSA.vue
Comment (richard-cox) [reply shell/components/LabelsPSA.vue:29]: The getter method can be added to the namespace model. This is the only place where `getPSALabels` is used

File: shell/components/LabelsPSA.vue
Comment (richard-cox) [reply shell/components/LabelsPSA.vue:1]: I've seen the design now, most of the conversation is more relevant there (given that we can hopefully use the existing `Labels` component

File: shell/components/ResourceDetail/index.vue
Comment (richard-cox) [reply shell/components/ResourceDetail/index.vue:381]: The same explanation given for keeping `DetailTop` generic apply here. There should be no type specific code in generic components. 

I can't see where these generic functions that rely on labels are used outside of namespaces, either here or inhttps://github.com/rancher/dashboard/pull/7678/files. Even if they are used in both they can still either use the same generic function or be in their on `BasePSA` class that the other classes inherit.

File: shell/components/LabelsPSA.vue
Comment (cnotv) [reply shell/components/LabelsPSA.vue:1]: This has been also discussed and we cannot use the existing `Labels` component, as this configuration is exceptional to Namespace only. Your option was already evaluated as first approach, but the component has nothing relevant to be reused and is rather complicating the view. I opted for keeping the tabs separated to avoid issues.

File: shell/components/DetailTop.vue
Comment (cnotv) [reply shell/components/DetailTop.vue:102]: The `value` computation from the model is already inside the component, which is not abstracted, I am not so sure to follow.
To avoid labels of certain types is part of the `value` computation like the others. Why is this an exception?

File: shell/components/ExplorerProjectsNamespaces.vue
Comment (cnotv) [reply shell/components/ExplorerProjectsNamespaces.vue:245]: As we have a parameter, I'm force to turn into a simple method. 

Screenshot 2022-12-12 at 15 45 04

Comment (cnotv) [reply shell/components/ExplorerProjectsNamespaces.vue:245]: As we have a parameter, I'm force to turn into a simple method. 

Screenshot 2022-12-12 at 15 45 04

File: shell/components/DetailTop.vue
Comment (richard-cox) [reply shell/components/DetailTop.vue:102]: `DetailTop` is a generic component, it has no reference to type specific code (PSALabelsNamespaceVersion) only generic methods exposed by models/classes. 

The model for namespaces can be updated to supply the type specific information that is accessed by the generic component in a generic way. Depending on how the Labels component works this can be the `get labels` getter in namespaces. If you still need the original labels list for other components, it looks like you can add a `get filteredSystemLabels` to namespaces that will contain the PSA label filtering

File: shell/components/ExplorerProjectsNamespaces.vue
Comment (richard-cox) [reply shell/components/ExplorerProjectsNamespaces.vue:245]: Sorry, typo above. The `row` is the instance available on the model. so it would be...

```
  get hasPSALabels() {
    return hasPSALabels(this); 
  }
```

File: shell/edit/namespace.vue
Comment (cnotv) [reply shell/edit/namespace.vue:230]: It's not the same, we are not using that logic somewhere else.

**Comment by cnotv** [comment]
@richard-cox It seems convenient to discuss about restructuring the whole architecture in [an issue](https://github.com/rancher/dashboard/issues/7684) where we can tackle down all the other generated problems caused by moving tests and TS inside the models. We will need a full perspective of all the use cases, so we do not omit or overcomplicate anything.

As this causes no harm, I'll opt to focus on the functionalities, e.g. solve the issue with the lack of Annotations from your new requirements.

It would be also constructive to [write a documentation with the enforced guidelines](https://github.com/rancher/dashboard/issues/7682) of all these points, so we do not have them dispersed. [The point from you mentioned do not seems present in the guidelines](https://rancher.github.io/dashboard/code-base-works/api-resources-and-schemas#models).

I have also created an issue where we could separately [make some analysis to eventual issues in adopting TS and unit tests](https://github.com/rancher/dashboard/issues/7683), so it may be simpler to point it out.

File: shell/components/ResourceDetail/index.vue
Comment (richard-cox) [reply shell/components/ResourceDetail/index.vue:381]: The three functions getPSALabels, hasPSALabels, getPSATooltipsDescription only apply to namespaces. PodSecurityAdmissionConfigurationTemplate work with a different object.

The description in question here only applies to namespaces.

Keeping resource type code out of generic components is what improves (or does not make worse) the code. There are no other cases where we use resource type code in generic components.

This does not prevent the use of TS or creating simple tests (models are easily created).

**Comment by cnotv** [comment]
@richard-cox ok.

File: shell/edit/namespace.vue
Comment (cnotv) [reply shell/edit/namespace.vue:230]: After talking with @kwwii there are the changes to be applied:
- We can merge the 2 labels components
- Correct the Label title size and space to match the tab
- Border to be removed
- Toggler can be hidden if no system labels
- Description is kept

This new style is applied to all the `Labels and Annotations` tabs, which means that not having system labels, it will not have toggler and no disabled inputs.

Comment (cnotv) [reply shell/edit/namespace.vue]: Updated and added screenshot.

**Comment by cnotv** [comment]
@richard-cox If you want to enforce the use of the models, please help me out with the migration to Typescript in https://github.com/rancher/dashboard/issues/7683 so I don't have to delete my work, as there were no issues as standalone implementation.
I am not able to override the key `metadata` inside the class Namespace, as it would not match with the inherited values. Any plan/suggestion?
You have also the PR assigned to you, so you can check what I did and the issue in re-assigning `metadata` breaking somewhere for some reasons: https://github.com/rancher/dashboard/pull/7726

**Comment by cnotv** [comment]
@richard-cox As alternative we may refer to the helpers from the classes, then when we complete [the model conversion to TS](https://github.com/rancher/dashboard/issues/7683) we can remove them.

File: shell/components/DetailTop.vue
Comment (cnotv) [reply shell/components/DetailTop.vue:102]: I'm trying to understand how to move the config into the model, but:
- We have already a config for the edit mode on top
- Labels is not a model, just a list from the value
- The value is not always namespace, so if I move the configuration it's not always present

What should I do in this case?

File: shell/components/LabelsPSA.vue
Comment (cnotv) [reply shell/components/LabelsPSA.vue]: Moved within model.

File: shell/components/ResourceDetail/index.vue
Comment (cnotv) [reply shell/components/ResourceDetail/index.vue:381]: Moved within model.

Comment (richard-cox) [reply shell/components/ResourceDetail/index.vue:383]: Resolved by https://github.com/cnotv/dashboard/pull/7

**Review by richard-cox — CHANGES_REQUESTED** [review]
I've resolved the remaining comments, fixed one bug and improved some visual elements in https://github.com/cnotv/dashboard/pull/7. Once that's merged to this branch's PR this can go in

File: shell/components/LabelsPSA.vue
Comment (richard-cox) [reply shell/components/LabelsPSA.vue]: Resolved by https://github.com/cnotv/dashboard/pull/7

**Comment by cnotv** [comment]
@richard-cox PR merged, still needs approval.

**Comment by cnotv** [comment]
I am getting this error after merging your PR @richard-cox 
Screenshot 2023-01-06 at 11 16 40

It seems that the `value` model does not have the method `getPSALabels()` on page loaded.
This happens for all the other labels which do not use the Namespace model.

**Comment by cnotv** [comment]
@richard-cox seen the bug fixed with the PSA labels. Forgot to pull the merged PR.

**Review by richard-cox — DISMISSED** [review]
I've pushed one final commit (see commit message for more details, but roughly... it fixes something related but not caused by this pr, removes usages of lodash and aligns sort order of PSAs in ns list and ns edit). 

If those look good to you then feel free to merge

**Comment by cnotv** [comment]
@richard-cox you still have to approve.

**Review by richard-cox — APPROVED** [review]

**Pull request merged**
